### PR TITLE
Validation and error types

### PR DESCRIPTION
**Pull request by BeeMargarida**
| - | - |
| --- | --- |
| Problem | Types were necessary for making model validation in typescript and errors. |
| Decisions | Types for validation functions and for error classes. |

### CODE REVIEW

**Comment by BeeMargarida** [comment]
@gcandal @joamag The PR we talked in the daily

File: types/base/error.d.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+export declare class YoniusError {
+    constructor(message: string, code?: number)
+    isClient() : Boolean
+    isServer() : Boolean
```
Comment (gcandal) [inline types/base/error.d.ts]:
```suggestion
    isClient(): boolean
    isServer(): boolean
```

File: types/util/validation.d.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+export declare function isUrl(
+    value?: unknown,
+    ctx?: any
+): Boolean;
+
+export declare function isRegex(
+    value?: unknown,
+    ctx?: any
+): Boolean;
```
Comment (gcandal) [inline types/util/validation.d.ts]: I believe these types are not correct. All these functions:
- Receive either 1 (message) or 2 (valueC, message) arguments
- Return a function that receive a (value, ctx) and return a boolean. The `value` should be typed

File: types/base/error.d.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    constructor(message?: string, code?: number)
+}
+
+export declare class NotImplementedError extends OperationalError {
+    constructor(message?: string, code?: number)
+}
+
+export declare class ValidationError extends OperationalError {
+    constructor(message?: string, code?: number)
+}
```
Comment (gcandal) [inline types/base/error.d.ts]: Is this needed?

**Review by gcandal — CHANGES_REQUESTED** [review]

File: types/base/error.d.ts (hunk 3 above)
Comment (BeeMargarida) [reply types/base/error.d.ts]: The constructor part, or the class declaration?

Comment (gcandal) [reply types/base/error.d.ts]: The constructor

**Comment by BeeMargarida** [comment]
@gcandal Re-request review is not working, just to remind and send notification of the fixes

File: types/base/error.d.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+export declare class YoniusError {
+    constructor(message: string, code?: number)
+    isClient() : boolean
+    isServer() : boolean
```
Comment (gcandal) [inline types/base/error.d.ts]: @BeeMargarida there's seems to be an extra space here. Are you the one adding it or prettier?
```suggestion
    isClient(): boolean
    isServer(): boolean
```

File: types/base/error.d.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+export declare class OperationalError extends YoniusError {
+    constructor(message?: string, code?: number)
```
Comment (gcandal) [inline types/base/error.d.ts:8]: Isn't the definition of the constructor inherited from the parent? If it is then this is not needed. If it's not, then we need to define the types of the constructors for all classes.

File: types/util/validation.d.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+): (value?: unknown, ctx?: any) => boolean;
+
+export declare function notEmpty(
+    message?: string
+): (value?: string | unknown[], ctx?: any) => boolean;
+
+export declare function isIn(
+    valueC: unknown[],
+    message?: string
+): (value?: unknown, ctx?: any) => boolean;
```
Comment (gcandal) [inline types/util/validation.d.ts]: Use generics for a stricter type enforcing.
```suggestion
export declare function eq<T>(
    valueC: T,
    message?: string
): (value?: T, ctx?: any) => boolean;

export declare function gt<T>(
    valueC: T,
    message?: string
): (value?: T, ctx?: any) => boolean;

export declare function notEmpty(
    message?: string
): (value?: string | unknown[], ctx?: any) => boolean;

export declare function isIn<T>(
    valueC: T[],
    message?: string
): (value?: T, ctx?: any) => boolean;
```

**Review by gcandal — CHANGES_REQUESTED** [review]

File: types/base/error.d.ts (hunk 5 above)
Comment (BeeMargarida) [reply types/base/error.d.ts:8]: I only left this one here since the message is now optional in this class

**Review by gcandal — APPROVED** [review]

**Comment by gcandal** [comment]
@joamag ready

**Review by joamag — APPROVED** [review]

**Pull request merged**